### DISCUSSION TITLE
Solo navmesh generation fails for big terrain

### DISCUSSION THREAD

**Discussion by yelouafi** [isaac-mason/recast-navigation-js#151] [General]
Trying to generate a solo navmesh with a 0.2 for the scene attached below;  fails with the error `Failed to build regions` or `Failed to create Detour navmesh data`; 

It succeeds though if I use 0.5 voxel size

[scene.glb](https://res.cloudinary.com/oncybertesting/raw/upload/v1695509656/1695509655-uploaded-assets/f74762827ce463f79b50d2acadf1e719499723ab040b33d57cf48bf2bd060e2d.glb)

**Comment by isaac-mason** [comment]
Thank you for reporting the issue and providing the model!

I believe I've reproduced this issue in Recast itself using [RecastDemo](https://github.com/recastnavigation/recastnavigation/tree/main#recastdemo). I'll need to investigate further, but we might need to:
a) Check if this is fixed in the latest Recast versions.
b) If not, consider upstreaming this issue.

RecastDemo also fails to generate a navmesh with cell size and cell height both set to 0.2 and a slope angle of 60. I'll need to confirm if this failure is for the same reason.

Interestingly, these variations also succeed:
- Using a "Walkable Slope Angle" of 45 instead of 60 (45 is RecastDemo's default)
- Using a cell size of 0.3 and a cell height of 0.2

<img width="1646" alt="image" src="https://github.com/isaac-mason/recast-navigation-js/assets/67411435/882d5630-6ac4-4414-b830-007f2ea6fda6">

<img width="1728" alt="image" src="https://github.com/isaac-mason/recast-navigation-js/assets/67411435/2c2a6b70-d6a7-4b02-946f-1d4a199f4c9d">

Here's the model as a obj, RecastDemo doesn't support gltf/glb - [big-terrain.obj.zip](https://github.com/isaac-mason/recast-navigation-js/files/12707543/big-terrain.obj.zip)

**Comment by isaac-mason** [comment]
Looking at the buildContext logs, this appears to be the root cause:

https://github.com/recastnavigation/recastnavigation/blob/cd898904b72a300011fbb24d578620bafa08ef2c/Recast/Source/RecastRegion.cpp#L1621

This looks like a limitation of the watershed partitioning method that `recast-navigation-js` uses for it's default solo nav mesh generator.

If you're interested, you can read more about the different partitioning options here: https://github.com/recastnavigation/recastnavigation/blob/cd898904b72a300011fbb24d578620bafa08ef2c/RecastDemo/Source/Sample_SoloMesh.cpp#L522

**Comment by isaac-mason** [comment]
The general advise I've read on recast's google discussions + github is to create tiled navmeshes for large maps.

That said, the tiled nav mesh generator also appears to have some bugs. Generation succeeds, but tiles aren't connected.

Let's at least leave this issue open until tiled nav mesh generation issues are fixed 🙂 

<img width="858" alt="image" src="https://github.com/isaac-mason/recast-navigation-js/assets/67411435/349ecc80-fc3d-498a-b197-467aa1480006">

**Comment by isaac-mason** [comment]
Fix for tiled nav mesh generation incoming 🙂 

<img width="860" alt="image" src="https://github.com/isaac-mason/recast-navigation-js/assets/67411435/c3b309b8-10e2-4dc8-a3bd-0fdc79ee7a6c">

**Comment by isaac-mason** [comment]
v0.10.2 will be released shortly with the tiled nav mesh generator fix, and navmesh.isaacmason.com has been updated now.

Let's convert this issue to a discussion, I think others will benefit from this.

LMK if you have any questions!

**Comment by yelouafi** [comment]
thanks for the quick answer! Yeah I was just looking into tiled navmeshes 

One thing I noticed is that with tiled generation the agent behaves sometimes unexpectdly, for example in this screenchot the agent seems stuck in its path and doesen't move forward

![Screenshot 2023-09-24 at 11 14 53](https://github.com/isaac-mason/recast-navigation-js/assets/5453835/02e7b8b8-44cf-4483-ae54-b4c49760d998)

**Reply by isaac-mason** [reply]
I've run into this issue too. Sometimes a fair bit of tweaking of both config and input is required to get a good output.

https://github.com/isaac-mason/recast-navigation-js/assets/67411435/c2dfaa02-dabb-4424-9e2e-6e61813e0586

I've noted that I usually get stuck agent issues like this when there are multiple layers / levels on the navmesh. Here it appears the agent is getting stuck right around the boundary of a lower layer.

https://github.com/isaac-mason/recast-navigation-js/assets/67411435/99bcdf52-3aeb-43a1-a5c1-61d41387b43d

A basic mitigation could be removing those lower non-visible triangles from the mesh passed into recast. 

I'll need to check if the recastnavigation google discussions or github issues have any wisdom on this 🙂

**Reply by isaac-mason** [reply]
Adding support for flood-fill navmesh pruning could help with problems like these. Created an issue: https://github.com/isaac-mason/recast-navigation-js/issues/156

**Reply by yelouafi** [reply]
ah I see the input mesh has some hidden triangles! I use the colliders to generate the navmesh and they shouldn't have this triangles. I'll check with our 3D modeler.

Thanks again for your help!

**Comment by yelouafi** [comment]
BTW the navmesh tiled generation fails with 0.1 cell size
